// session.socket.io/session.socket.io.js
module.exports = SessionSockets;

function SessionSockets(io, sessionStore, cookieParser, key) {
  key = typeof key === 'undefined' ? 'connect.sid' : key;
  var sessionSockets = this;
  this.io = io;

  this.on = function(event, callback) {
    return bind(event, callback, io.sockets);
  };

  this.of = function(namespace) {
    return {
      on: function(event, callback) {
        return bind(event, callback, io.of(namespace));
      }
    };
  };

/*
  this.getSession = function(socket, callback) {
    cookieParser(socket.handshake, {}, function (parseErr) {
      var sessionLookupMethod = sessionStore.load || sessionStore.get;
      sessionLookupMethod.call(sessionStore, findCookie(socket.handshake), function (storeErr, session) {
        var err = resolveErr(parseErr, storeErr, session);
        console.log(err);
        console.log(session);
        callback(err, session, socket);
      });
    });
  };
*/
  this.getSession = function(socket, callback) {
    cookieParser(socket.handshake, {}, function (parseErr) {
      var sessionid = socket.handshake.cookies[key];
      //由于cookie保存如下，mongodb保存的_id为sessionID的加密格式，为ypbxGhOzzAQDlsn3mmVGrO6A，获取想要的串就ok
      //s:ypbxGhOzzAQDlsn3mmVGrO6A.9rtRxhQu3r8ymOZ1s%2FzlfEWaZ9MSorSBobzy8CNhjh0
      var sid = '';
      if(sessionid)
      {
        sid =  sessionid.split(':')[1].split('.')[0];
      }

      sessionStore.get(sid, function (storeErr, session) {
        var err = resolveErr(parseErr, storeErr, session);
        callback(err, session, socket);
      });
    });
  };

  function bind(event, callback, namespace) {
    namespace.on(event, function (socket) {
      sessionSockets.getSession(socket, function (err, session) {
        callback(err, socket, session);
      });
    });
  }

  function findCookie(handshake) {
    if (handshake)
      return (handshake.secureCookies && handshake.secureCookies[key]) ||
             (handshake.signedCookies && handshake.signedCookies[key]) ||
             (handshake.cookies && handshake.cookies[key]);
  }

  function resolveErr(parseErr, storeErr, session) {
    var err = parseErr || storeErr || null;
    if (!err && !session) err = new Error('Could not lookup session by key: ' + key);
    return err;
  }
}
